ethereum/contracts/relayer/coreRelayer/CoreRelayerGovernance.sol

```diff
@@ -26,76 +26,84 @@ error InvalidFork();
+   *   - https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/BridgeGovernance.sol#L115
+   */
+
+  /**
+   * Registers a core relayer contract that was deployed on another chain with the CoreRelayer on
+   *   this chain. The equivalent to the core bridge's registerChain action.
+   *
+   * Action Parameters:
+   *   - uint16 foreignChainId - originally emitterChainID, renamed for clarity
+   *   - bytes32 foreignContractAddress - originally emitterAddress, renamed for clarity
+   */
+  uint8 private constant GOVERNANCE_ACTION_REGISTER_CORE_RELAYER_CONTRACT = 1;
+
+  /**
+   * Upgrades the CoreRelayer contract to a new implementation. The equivalent to the core bridge's
+   *   upgrade action.
+   *
+   * Action Parameters:
+   *   - bytes32 newImplementation - bytes32 instead of address for general consistency reasons
+   */
+  uint8 private constant GOVERNANCE_ACTION_CONTRACT_UPGRADE = 2;
+
+  /**
+   * Sets the default relay provider for the CoreRelayer. Has no equivalent in the core bridge.
+   *
+   * Action Parameters:
+   *   - bytes32 newProvider - bytes32 instead of address for general consistency reasons
+   */
+  uint8 private constant GOVERNANCE_ACTION_UPDATE_DEFAULT_PROVIDER = 4;
+
+  //By checking that only the contract can call itself, we can enforce that the migration code is
+  //  executed upon program upgrade and that it can't be called externally by anyone else.
+  function checkAndExecuteUpgradeMigration() external {
+    assert(msg.sender == address(this));
+    executeUpgradeMigration();
+  }
+
+  function executeUpgradeMigration() internal virtual {
+    //override and implement in CoreRelayer upon contract upgrade (if required)
+  }
+
+  function registerCoreRelayerContract(bytes memory encodedVm) external {
+    (uint16 foreignChainId, bytes32 foreignAddress) =
+      parseAndCheckRegisterCoreRelayerContractVm(encodedVm);
+
+    getRegisteredCoreRelayersState().registeredCoreRelayers[foreignChainId] = foreignAddress;
+  }
+
   function submitContractUpgrade(bytes memory encodedVm) external {
     address newImplementation = parseAndCheckContractUpgradeVm(encodedVm);
 
     _upgradeTo(newImplementation);
 
-    //Now that our implementation points to the new contract, we can just call ourselves and have
-    //  executeUpgradeMigration() use a simple check that msg.sender == address(this).
-    //I.e. the new contract just has to implement:
-    //  function executeUpgradeMigration() external view {
-    //    assert(msg.sender == address(this));
-    //    //migration code goes here
-    //  }
-    //  so we don't have to mess around with any "intialized" and "reinitialized" storage vars.
     (bool success, bytes memory revertData) =
-      address(this).call(abi.encodeWithSignature("executeUpgradeMigration()"));
+      address(this).call(abi.encodeWithSelector(this.checkAndExecuteUpgradeMigration.selector));
```

Comment:
Now that we can use the latest compiler version we can just go with the typechecked version:
```suggestion
      address(this).call(abi.encodeCall(this.checkAndExecuteUpgradeMigration, ()));
```

Reply:
good call (pun intended) :laughing: